File: backend/exam-service/internal/models/exam.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type ExamStatus string

const (
	ExamStatusDraft     ExamStatus = "draft"
	ExamStatusPublished ExamStatus = "published"
	ExamStatusOngoing   ExamStatus = "ongoing"
	ExamStatusEnded     ExamStatus = "ended"
	ExamStatusGraded    ExamStatus = "graded"
)

type ExamType string

const (
	ExamTypeQuiz       ExamType = "quiz"
	ExamTypeMidterm    ExamType = "midterm"
	ExamTypeFinal      ExamType = "final"
	ExamTypeMock       ExamType = "mock"
	ExamTypePractice   ExamType = "practice"
	ExamTypeAssignment ExamType = "assignment"
)

type Exam struct {
	ID          uuid.UUID  `json:"id" db:"id"`
	Name        string     `json:"name" db:"name"`
	Type        ExamType   `json:"type" db:"type"`
	SubjectIDs  []string   `json:"subjectIds" db:"subject_ids"`
	ClassIDs    []string   `json:"classIds" db:"class_ids"`
	StartTime   time.Time  `json:"startTime" db:"start_time"`
	EndTime     time.Time  `json:"endTime" db:"end_time"`
	Duration    int        `json:"duration" db:"duration"`
	Status      ExamStatus `json:"status" db:"status"`
	TotalScore  float64    `json:"totalScore" db:"total_score"`
	PassScore   float64    `json:"passScore" db:"pass_score"`
	Description string     `json:"description" db:"description"`
	CreatedBy   uuid.UUID  `json:"createdBy" db:"created_by"`
	CreatedAt   time.Time  `json:"createdAt" db:"created_at"`
	UpdatedAt   time.Time  `json:"updatedAt" db:"updated_at"`
}

func NewExam() *Exam {
	return &Exam{
		ID:        uuid.New(),
		Status:    ExamStatusDraft,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}
}

type ExamCreateRequest struct {
	Name        string   `json:"name" binding:"required"`
	Type        ExamType `json:"type" binding:"required"`
	SubjectIDs  []string `json:"subjectIds" binding:"required"`
	ClassIDs    []string `json:"classIds" binding:"required"`
	StartTime   string   `json:"startTime" binding:"required"`
	EndTime     string   `json:"endTime" binding:"required"`
	Duration    int      `json:"duration" binding:"required,min=1"`
	TotalScore  float64  `json:"totalScore"`
	PassScore   float64  `json:"passScore"`
	Description string   `json:"description"`
}

type ExamUpdateRequest struct {
	Name        string   `json:"name"`
	Type        ExamType `json:"type"`
	SubjectIDs  []string `json:"subjectIds"`
	ClassIDs    []string `json:"classIds"`
	StartTime   string   `json:"startTime"`
	EndTime     string   `json:"endTime"`
	Duration    int      `json:"duration"`
	TotalScore  float64  `json:"totalScore"`
	PassScore   float64  `json:"passScore"`
	Description string   `json:"description"`
}

type ExamListResponse struct {
	Exams      []Exam `json:"exams"`
	Total      int64  `json:"total"`
	Page       int    `json:"page"`
	PageSize   int    `json:"pageSize"`
	TotalPages int    `json:"totalPages"`
}

type ExamFilter struct {
	Status    ExamStatus `form:"status"`
	Type      ExamType   `form:"type"`
	SubjectID string     `form:"subjectId"`
	ClassID   string     `form:"classId"`
	StartDate string     `form:"startDate"`
	EndDate   string     `form:"endDate"`
	Page      int        `form:"page" binding:"min=1"`
	PageSize  int        `form:"pageSize" binding:"min=1,max=100"`
}

File: backend/grade-service/app/schemas/exam.py
from pydantic import BaseModel, Field, ConfigDict
from datetime import datetime
from uuid import UUID
from typing import Optional, List
from enum import Enum


class ExamType(str, Enum):
    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"
    MIDTERM = "midterm"
    FINAL = "final"
    MOCK = "mock"


class ExamStatus(str, Enum):
    DRAFT = "draft"
    SCHEDULED = "scheduled"
    ONGOING = "ongoing"
    COMPLETED = "completed"
    PUBLISHED = "published"


class ExamBase(BaseModel):
    name: str = Field(..., max_length=200)
    type: ExamType = ExamType.DAILY
    start_time: datetime
    end_time: datetime
    grade_level: Optional[str] = Field(None, max_length=50)
    semester: Optional[str] = Field(None, max_length=50)
    school_id: Optional[UUID] = None
    description: Optional[str] = None
    total_score: Optional[float] = None


class ExamCreate(ExamBase):
    pass


class ExamUpdate(BaseModel):
    name: Optional[str] = Field(None, max_length=200)
    type: Optional[ExamType] = None
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    status: Optional[ExamStatus] = None
    grade_level: Optional[str] = Field(None, max_length=50)
    semester: Optional[str] = Field(None, max_length=50)
    description: Optional[str] = None
    total_score: Optional[float] = None


class ExamResponse(ExamBase):
    id: UUID
    status: ExamStatus = ExamStatus.DRAFT
    created_at: datetime
    updated_at: datetime
    
    model_config = ConfigDict(from_attributes=True)


class ExamListResponse(BaseModel):
    items: List[ExamResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class ExamFilter(BaseModel):
    type: Optional[ExamType] = None
    status: Optional[ExamStatus] = None
    grade_level: Optional[str] = None
    semester: Optional[str] = None
    school_id: Optional[UUID] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1, le=100)
